fix(sql): fix file does not exist error on updates

ideoma opened the pull request:
Fix column top handling in UPDATE execution

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 28 / 28 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/WalPurgeJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/ConvertOperatorImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoException.java|2|2|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 29 / 29 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/WalPurgeJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/CopyWalSegmentUtils.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/ConvertOperatorImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoException.java|2|2|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 36 / 37 (97.30%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/CopyWalSegmentUtils.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalPurgeJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/SegmentColumnRollSink.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/ConvertOperatorImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoException.java|2|2|100.00%|

puzpuzpuz commented on core/src/main/java/io/questdb/cairo/TableUtils.java:
```diff
@@ -1149,7 +1149,11 @@ public static int openRO(FilesFacade ff, LPSZ path, Log log) {
-        throw CairoException.critical(ff.errno()).put("could not open read-only [file=").put(path).put(']');
+        int errno = ff.errno();
+        if (CairoException.errnoReadPathDoesNotExist(errno)) {
+            throw CairoException.critical(errno).put("could not open, file does not exists: ").put(path).put(']');
```
nit: typo here
```suggestion
            throw CairoException.critical(errno).put("could not open, file does not exist: ").put(path).put(']');
```

ideoma replied:
done

puzpuzpuz commented on core/src/main/java/io/questdb/cairo/TableWriter.java:
```diff
@@ -538,7 +538,7 @@ public void addColumn(
 
         long columnNameTxn = getTxn();
-        LOG.info().$("adding column '").utf8(columnName).$('[').$(ColumnType.nameOf(columnType)).$("], columnName txn ").$(columnNameTxn).$(" to ").$substr(pathSize, path).$();
+        LOG.info().$("adding column '").utf8(columnName).$('[').$(ColumnType.nameOf(columnType)).$("], columnName txn ").$(columnNameTxn).$(" to ").$substr(pathRootSize, path).$();
```
nit: would be nice to make path logging consistent, e.g. in `addIndex()` it's `.$(path)`, not `.$substr(pathRootSize, path)`

ideoma replied:
I changed the place you mentioned and a few more. I keep full path on partition manipulation and error logging

puzpuzpuz commented on core/src/test/java/io/questdb/test/griffin/wal/WalWriterFuzzTest.java:
```diff
@@ -167,6 +169,27 @@ public void testWalMetadataChangeHeavy() throws Exception {
     }
 
+    @Test
+    public void testWalMetadataAddDeleteColumHeavy() throws Exception {
```
nit:
```suggestion
    public void testWalMetadataAddDeleteColumnHeavy() throws Exception {
```

ideoma replied:
done

puzpuzpuz commented on core/src/test/java/io/questdb/test/griffin/UpdateFuzzTest.java:
```diff
@@ -0,0 +1,163 @@
+                }
+            }
+
+            printSql("up");
```
nit: do we need this query execution considering that we don't do any later assertions on the sink?

ideoma replied:
I added the comment 

```

            // This will read whole table, verifying that the TableReader can handle the column top / versions / types
            printSql("up");
```

puzpuzpuz replied:
Thanks!

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 45 / 48 (93.75%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/CopyWalSegmentUtils.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|10|12|83.33%|
|:large_blue_circle:|io/questdb/cairo/wal/WalPurgeJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/SegmentColumnRollSink.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/ConvertOperatorImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoException.java|2|2|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 45 / 48 (93.75%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/CopyWalSegmentUtils.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|10|12|83.33%|
|:large_blue_circle:|io/questdb/cairo/wal/WalPurgeJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/SegmentColumnRollSink.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/ConvertOperatorImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoException.java|2|2|100.00%|

puzpuzpuz commented:
Hmm, `ServerMainBackupDatabaseTest.testServerMainBackupDatabase` failed in multiple CI runs

puzpuzpuz commented:
`ServerMainBackupDatabaseTest.testServerMainBackupDatabase` failed twice on `mac-other`. @ideoma could you check if that's flakiness or not?

ideoma commented:
> `ServerMainBackupDatabaseTest.testServerMainBackupDatabase` failed twice on `mac-other`. @ideoma could you check if that's flakiness or not?

in a way, because I changed a random generation, now the setup decides to create WAL table that never happened before

```Java
    @Before
    public void setUp() {
        super.setUp();

        isWal = rnd.nextBoolean();
...
```

I fixed the test handle WAL tables properly at https://github.com/questdb/questdb/pull/4846/commits/7f548fb736d4b0cdbc837c318820b157f9e9ec9e

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 45 / 48 (93.75%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/CopyWalSegmentUtils.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|10|12|83.33%|
|:large_blue_circle:|io/questdb/cairo/wal/WalPurgeJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/SegmentColumnRollSink.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/ConvertOperatorImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoException.java|2|2|100.00%|
